_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,381 @@
+```ruby
+Polygon.new(___).perimeter
+```
+> In Ruby, and in most OOP languages, a `class` describes a "template" for an object. An object is usually a collection of data, and a set of methods or functions that allow you to interact with that data or give it logic. We first describe the objects with a `class`, and create new instances of them with `Thing.new`.
+
+In the next sections, we'll work with one approach to solve this from the ground up, to figure out how to *make* a `Polygon` to get what we need, and explore some properties of Ruby along the way.
+
+## Objects
+
+### Point
+
+Consider any kind of object around you. A pen, a car, your computer - they all have properties, or *attributes*.  Ruby objects can be given *attributes* that let you describe and store data inside them. We do this with the `@var` syntax, like so, describing what defines a `Point` in 2D space:
+
+```ruby
+# A location in 2D space.
+class Point
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> `def` defines a new method, or an instruction on how to do something with an object. `initialize` is a special method that is called whenever you call `#new` on an object to make a new instance of one.
+
+We can now make a new `Point` like so:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.inspect #=> #<Point:0x2ae37f0 @x=1.5, @y=-3.0>
+```
+
+We're going to need to use these attributes of a point in order to do our calculations. To do so, we can define methods that will return the values of our attributes:
+
+```ruby
+# A location in 2D space.
+class Point
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+
+  def x
+    @x
+  end
+
+  def y
+    @y
+  end
+end
+```
+Since this is such a common thing in Ruby, the language has shortcuts to quickly define such simple methods that reflect an objects attributes:
+
+- `attr_reader :var` - Creates a read-only method for attribute `@var`.
+- `attr_writer :var` - Creates a write-only method for attribute `@var`.
+- `attr_accessor :var` - Creates both kinds of methods to read and write to this attribute.
+
+So, our `Point` can be described simply as:
+```ruby
+# A location in 2D space.
+class Point
+  attr_reader :x
+
+  attr_reader :y
+
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> In our case, we will only need to ever read these attributes, so I just used `attr_reader`. `attr_accessor` would have also worked here.
+
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
+
+Ruby has a rich standard library, of which includes mathematical functions. Needing the pythagorean theorem here, if we do some searching we can see Ruby already provides a method for us to calculate this instead of having to implement it ourselves:
```

Comment:
"Needing the pythagorean theorem here, if we do some searching we can see ..."
Kinda tangled my brain a little. Something like "We need the Pythagorean theorem to calculate the distance between two points, and doing some searching we can see ..." would work better